### DISCUSSION TITLE
"Requested function was not found" but invocations are logged

### DISCUSSION THREAD

**Discussion by glennslaven** [supabase/supabase#26818] [Questions]
I've got an edge function that worked fine in local when served the CLI, but once I deploy it to supabase hosted it just returns
```{"code":"NOT_FOUND","message":"Requested function was not found"}```
However, when I look at the function in the portal, the invocations are there, showing they return a 404. What am I missing? I've double checked and nowhere in my function code do I return a 404, and if I put a `console.log` at the start of the function, that log message never shows up in the logs.

For reference, this is the code of the function:

```ts
import pdf from "npm:pdf-parse/lib/pdf-parse.js";

Deno.serve(async (req) => {
  try {
    const reqFormData = await req.formData();

    if (!reqFormData.has("file")) {
      throw new Error("Missing required 'file' property from request data.");
    }

    const file = reqFormData.get("file") as File;
    const data = await pdf(await file.arrayBuffer());
    if (!data) throw new Error("Error parsing PDF");

    return new Response(JSON.stringify(data.text), {
      headers: { "Content-Type": "application/json" },
    });
  } catch (err) {
    console.error(err);
    return new Response(JSON.stringify(err), {
      status: 500,
      headers: { "Content-Type": "application/json" },
    });
  }
});
```

**Comment by glennslaven** [comment]
So as an experiment I tried doing a `supabase functions download` of this function and it threw a 500 error after a while. It spat out a cloudflare 500 error page to the console. Downloading my other functions work fine.

I also tried deleting and re-deploying the function, same result.

**Comment by glennslaven** [comment]
Ok, so this seems to be related to an npm package I'm using. I'm trying to parse a PDF file to extract the text, and I'm using pdf-parse as per https://stackoverflow.com/questions/76249978/is-there-a-way-to-fetch-pdf-from-url-and-extract-text-from-it-in-deno & it works in local but not when deployed to supabase.

If I try and import the whole pdf-parse package, ie `import pdf from 'npm:pdf-parse` it errors with 
```runtime has escaped from the event loop unexpectedly: event loop error: NotFound: path not found test/data/05-versions-space.pdf: readfile './test/data/05-versions-space.pdf'```

**Comment by tobiasr** [comment]
I have the same issue. I'm not using the pdf-parse package (I tried but didn't get it to work because of the ...space.pdf error. I'm using https://www.npmjs.com/package/pdf-ts instead.

But I'm getting the same `{"code":"NOT_FOUND","message":"Requested function was not found"}` when invoking it. It works fine locally.

**Comment by glennslaven** [comment]
So I put the function into [Deno Deploy playground](https://dash.deno.com/playground/expensive-fly-75) and it works fine there.

```bash
curl -F "file=@path-to-file.pdf" https://expensive-fly-75.deno.dev
```

Which is good, because it means the issue isn't just "Deno don't work like this, sorry", but it does mean that the issue is in _Supabase Flavored Deno<sup>TM</sup>_

**Comment by glennslaven** [comment]
Ok, so I have **"a"** solution. I thought I'd strip the code back and use [pdfjs](https://mozilla.github.io/pdf.js/) since that's what pdf-parse is using under the hood anyway, but doing an npm import of `pdfjs-dist` and rewriting the code to use that resulted in exactly the same outcome ie. a 404, function not found in the hosted edge function.

However, if I did an npm install of `pdfjs-dist`, copied out the `pdf.min.js` and `pdf.worker.min.js` files from the node_modules directory, stuck them in my function directory and imported them locally it actually works. 

Here's what the function looks like now:
```ts
import * as pdfjs from "./pdfjs/pdf.min.mjs";
import * as pdfjsworker from "./pdfjs/pdf.worker.min.mjs";

// Just a reference to force the import of the worker, otherwise it will not be included in the bundle
type p = typeof pdfjsworker;

pdfjs.GlobalWorkerOptions.workerSrc = "./pdf.worker.min.mjs";

Deno.serve(async (req) => {
  try {
    const reqFormData = await req.formData();

    if (!reqFormData.has("file")) {
      throw new Error("Missing required 'file' property from request data.");
    }

    const file = reqFormData.get("file") as File;

    const task = pdfjs.getDocument({ data: await file.arrayBuffer() });

    const data = await task.promise;
    if (!data) throw new Error("Error parsing PDF");

    let output = "";
    for (let i = 1; i <= data.numPages; i++) {
      const page = await data.getPage(i);
      const textContent = await page.getTextContent({ includeMarkedContent: false });
      const text = textContent.items.map((item: { str: string }) => item.str).join("\n");
      output += text;
    }

    return new Response(output, {
      headers: { "Content-Type": "application/json" },
    });
  } catch (err) {
    console.error(err);
    return new Response(JSON.stringify(err), {
      status: 500,
      headers: { "Content-Type": "application/json" },
    });
  }
});
```

You don't get all the nice typescript types on the pdfjs functions, everything's just `any`. But it works! And as a bonus it's only 441.2kB in size :)

Something is still wrong with this whole situation. My best guess is it's something to do with the filesystem restrictions that _Supabase Flavored Deno<sup>TM</sup>_ has, eg `Deno.readDir` doesn't exist inside edge functions. Just a guess though.

**Reply by limdingwen** [reply]
You are a *lifesaver*! Thank you!

Here's the same code in function form, for extracting from a URL:

```
import * as pdfjs from "../lib/pdfjs/pdf.mjs";
import * as pdfjsworker from "../lib/pdfjs/pdf.worker.mjs";

// Just a reference to force the import of the worker, otherwise it will not be included in the bundle
type p = typeof pdfjsworker;
pdfjs.GlobalWorkerOptions.workerSrc = "./pdf.worker.mjs";

async function extractTextFromPDF(pdfUrl: string) {
  const response = await fetch(pdfUrl);

  const task = pdfjs.getDocument({ data: await response.arrayBuffer() });
  const data = await task.promise;
  if (!data) throw new Error("Error parsing PDF");

  let output = "";
  for (let i = 1; i <= data.numPages; i++) {
    const page = await data.getPage(i);
    const textContent = await page.getTextContent({
      includeMarkedContent: false,
    });
    const text = textContent.items
      .map((item: { str: string }) => item.str)
      .join("\n");
    output += text;
  }

  return output;
}
```

Note that `.min` is not there in my code. I downloaded the bundle from https://mozilla.github.io/pdf.js/.

**Reply by matiasbattocchia** [reply]
You made my day! Thanks!

**Reply by olegkorol** [reply]
Thanks for the workaround, @glennslaven!
I've spent a great deal of time trying to debug the exact same issue... 😄

Here are the imports from `jsdelivr` in case someone wants to use them instead of downloading the files manually to a local folder:
```typescript
import * as pdfjs from 'https://cdn.jsdelivr.net/npm/pdfjs-dist@4.6.82/build/pdf.min.mjs'
import * as pdfjsworker from 'https://cdn.jsdelivr.net/npm/pdfjs-dist@4.6.82/build/pdf.worker.min.mjs'
```

**Comment by laktek** [comment]
@glennslaven @tobiasr When you deploy the function via CLI, what's the reported bundle size? My suspicion is that the function bundle is borderline near 20MB, and it fails to store properly in our data store. It should have a given a better error message at the time of deploy. I'll investigate why it didn't.

**Comment by glennslaven** [comment]
Mine is 14.67MB using pdf-parse or 12.37MB using pdfjs-dist. Both fail the same way with the 404. Using the above workaround it's only 441.2kB and that one works.

I just noticed on the [pricing page](https://supabase.com/pricing) that the "script size" max is 10MB. It's not clear to me what "script size" means, is that just my code, or does it also include the npm packages?

**Comment by laktek** [comment]
Yeah, advertised limit is 10MB (though technically, it should be possible up to 20MB). But from your example it seems it has issues serving functions above 10MB.

We have some improvements ready on our end which should help with reducing bundle size (eg: https://github.com/supabase/edge-runtime/pull/343). These changes should go live within the next week or two.

**Comment by JEndler** [comment]
Am having the same problem with a function that uses the `googleapis` npm package (which is >100mb big when unpacked 🙃 ).

What would be the "intended" way of dealing with this? I've tried to import only part of the package (eg. `npm:googleapis@140.0.0/oauth2`) but that didn't work.

**Reply by JEndler** [reply]
I managed to get it down to 12.6 MB by importing a googleapis Submodule: `npm:googleapis@140.0.0/dataportability`

Tho that still fails with a "Not Found".

I'm also not really sure where all that size comes from, I'm just importing the supabase client, and the google-data-portability-api (which, [according to npm](https://www.npmjs.com/package/@googleapis/dataportability), should be 103kb unpacked).

**Reply by JEndler** [reply]
I'm not really sure how to proceed here. Its not an option to import anything less.

I guess I'll try to move to Deno deploy for now.

**Comment by geoffreygarrett** [comment]
Facing a similar issue. Locally when I deploy my 29 MB function (about to go inspect what's causing the bloat) I get no feedback from the CLI about it failing. However, on GitHub actions I do, perhaps a version difference.

Downloading yields and equally non-informative result..

```bash
(base) ➜  nsri-link git:(develop) ✗ supabase functions download task
Downloading task
Error status 500: <!DOCTYPE html>         
```

**Reply by geoffreygarrett** [reply]
Complements of GPT, found my culprit.

```bash
deno info --import-map ./import_map.json task/index.ts | gawk '
{
    # Normalize input and directly convert and accumulate sizes
    gsub(/\x1B\[[0-9;]*[mK]/, "");  # Remove ANSI codes
    if (match($0, /(\S+)\s+\(([0-9]+\.[0-9]+)(KB|MB)\)/, arr)) {
        size = arr[2];
        if (arr[3] == "KB") size /= 1024;  # Convert KB to MB
        sizes[arr[1]] += size;  # Aggregate sizes by path
    }
}

END {
    # Sort and output the top 10 results
    asorti(sizes, sorted, "@val_num_desc");
    for (i = 1; i <= 10 && i <= length(sorted); i++)
        printf("%.2f MB\t%s\n", sizes[sorted[i]], sorted[i]);
}'

42.19 MB        npm:/gpt-tokenizer@2.1.2
6.27 MB npm:/js-tiktoken@1.0.10
4.38 MB npm:/html-minifier-terser@7.2.0
4.18 MB npm:/terser@5.31.0
3.51 MB npm:/@types/node@18.16.19
3.04 MB npm:/@putout/minify@4.3.0
1.91 MB npm:/grammy@1.22.4
1.59 MB npm:/@langchain/core@0.1.50
1.54 MB npm:/source-map@0.6.1
1.20 MB npm:/zod@3.22.4
```

**Comment by mojtabaasadi** [comment]
I am facing the same issue . although my script size is less than 20mb and my deployment seems to e successful.
```
Deploying k*** (script size: 17.22MB)
Deployed Function k*** on project ***
You can inspect your deployment in the Dashboard: https://supabase.com/dashboard/project/***/functions/k***/details
``` 

```
{
  "code": "NOT_FOUND",
  "message": "Requested function was not found"
}
```

**Reply by geoffreygarrett** [reply]
When I encountered the issue the deployment also seemed to be successful. Try this:

```bash
supabase functions download <k***>
```

And you should too receive:
```
Error status 500: <!DOCTYPE html>    
```

**Reply by geoffreygarrett** [reply]
If you use the gawk command above from a few weeks back, you'll quickly see which packages you should try find a lighter-weight alternative for. Often enough, you'll find a drop-in replacement for a large npm package specifically for deno at a fraction of the size.
